### ISSUE TITLE
localstorage and compression

### ISSUE THREAD

**Issue by vieridipaola** [InnovateAsterisk/Browser-Phone#86]
Hi,

Currently, I am storing the json buddy lists for all my users on a server. Some lists can be 100KB+ so I compress them client-side, store them on the server, then eventually download them and uncompress them so they can be imported in localstorage.

Since there are a lot of users, this compression saves bandwidth usage.

I believe localstorage has a size limit, am I right?

If so, wouldn't it be better to store the buddy list as "compressed" data with a lib such as the one in the following link?

https://pieroxy.net/blog/pages/lz-string/index.html
(eg. localDB.setItem(profileUserID + "-Buddies", LZString.compress(myBuddyList));  )

This is just a suggestion to improve performance in bug setups, and it shouldn't be at the top of your priority list...

Thanks

**Comment by InnovateAsterisk** [comment]
Yes, the local storage does have a limitation, but probably a bigger concern is that it's always synchronous.

http://jsfiddle.net/53xcc4LL/
It seems to be about 5000KB (5MB), so you are still quite safe.

When information is small (or few), then it works amazingly well, but yes when there are thousands of entries, I guess it isn't a good scale model. I'm not as worried about the size in bytes of the buddies database, it's mainly about the UI responsiveness and general performance. The IndexDB is a system that can expand to (apparently) half your system disk space is designed for truly massive amounts of data, is accessed asynchronous, and as the name says, the data can be indexed.

There is a TODO for this.

**Comment by vieridipaola** [comment]
OK, but what worries me more than the disk space on the client or server is the amount of data transferred between client and server when provisioning or saving (mostly the buddy list, in my case). Without compression, the average user's buddy list is around 100KB and growing. For now, the UI responsiveness is fine, so I haven't set quotas yet.  Also, the number of users on Browser-Phone is also growing, so the net result is that whenever there's a new buddy or a contact has been modified, the client sends "the whole list" to the server (I haven't had the time to find the way to detect changes only -- it's so much easier to just send everything ;-) ). So, if I foresee around 200 users mid-term and maybe 800 long-term, each with min. 100KB, that would translate to a lot of network traffic.

The lz-string lib I mentioned seems to compress 5x aprox.

Anyway, it will probably be a different story with indexDB so I guess we can put this on hold until then.

Thanks!

**Comment by InnovateAsterisk** [comment]
Ah.. .ok, is see your issue, but it's not really the way the system was designed... well the fact that you have the entire list download each time the page displays. Hmm let me think about that a bit - the best solution would be that once the "initial list" loads, its then up to the user to maintain their own list.

**Comment by vieridipaola** [comment]
1) a brand new user logs in for the first time: empty buddy list. He/she can decide to download or not a corporate buddy list (according to login details). That's up to the user.

2) When the user receives calls or "dials by line" then a buddy is automatically created, as expected.

3) The user can add a buddy either manually or import it from an intranet directory (I integrated the Browser-Phone with the Intranet).

4) There is a "setInterval function" that does the following:
a) it periodically sends the client buddy list to the server ONLY if the buddy list array is different from the previous in-memory copy,
b) it compresses the json string before sending it to the server

5) When the same user logs into the Browser-Phone page but on another browser/device, when the document is "ready" it queries the server for the compressed buddy list only once. It is uncompressed and saved as the localstorage buddy list then UpdateBuddyList is called, etc.. That allows a user to move around and keep his/her buddy list updated.
